runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonEncoder.kt

```diff
@@ -0,0 +1,157 @@
+
+    override fun beginObject() = openStructure("{", LexerState.ObjectFirstKeyOrEnd)
+    override fun endObject() = closeStructure("}")
+    override fun beginArray() = openStructure("[", LexerState.ArrayFirstValueOrEnd)
+    override fun endArray() = closeStructure("]")
+
+    private fun openStructure(token: String, nextState: LexerState) {
+        encodeValue(token)
+        if (pretty) writeNewline()
+        depth++
+        state.push(nextState)
+    }
+
+    private fun closeStructure(token: String) {
+        if (pretty) {
+            writeNewline()
+        }
+        depth--
+        writeIndent()
+        buffer.append(token)
+        state.pop()
+    }
+
+    private fun writeIndent() {
+        if (pretty && depth > 0) {
+            val indent = " ".repeat(depth * 4)
+            buffer.append(indent)
+        }
+    }
+
+    private fun writeNewline() = buffer.append("\n")
+    private fun writeComma() {
+        buffer.append(",")
+        if (pretty) writeNewline()
+    }
+
+    private fun writeColon() {
+        buffer.append(":")
+        if (pretty) buffer.append(" ")
+    }
+
+    private fun encodeValue(value: String) {
+        when (state.top()) {
+            LexerState.ArrayFirstValueOrEnd -> {
+                state.replaceTop(LexerState.ArrayNextValueOrEnd)
+                writeIndent()
+            }
+            LexerState.ArrayNextValueOrEnd -> {
+                writeComma()
+                writeIndent()
+            }
+            LexerState.ObjectFieldValue -> {
+                writeColon()
+                state.replaceTop(LexerState.ObjectNextKeyOrEnd)
+            }
+            else -> {}
+        }
+
+        buffer.append(value)
+    }
+
+    override fun writeNull() = encodeValue("null")
+
+    private fun StringBuilder.appendQuoted(value: String) {
+        append("\"")
+        append(value)
+        append("\"")
+    }
+
+    override fun writeName(name: String) {
+        if (state.top() == LexerState.ObjectNextKeyOrEnd) {
+            writeComma()
+        }
+        writeIndent()
+        buffer.appendQuoted(name.escape())
+        state.replaceTop(LexerState.ObjectFieldValue)
+    }
+
+    override fun writeValue(value: String) = encodeValue("\"${value.escape()}\"")
+
+    override fun writeValue(bool: Boolean) = when (bool) {
+        true -> encodeValue("true")
+        false -> encodeValue("false")
+    }
+
+    private fun writeNumber(value: Number) = encodeValue(value.toString())
+
+    override fun writeValue(value: Byte) = writeNumber(value)
+    override fun writeValue(value: Long) = writeNumber(value)
+    override fun writeValue(value: Short) = writeNumber(value)
+    override fun writeValue(value: Int) = writeNumber(value)
+    override fun writeValue(value: Float) = writeNumber(value)
+    override fun writeValue(value: Double) = writeNumber(value)
+
+    override fun writeRawValue(value: String) = encodeValue(value)
+}
+
+internal fun String.escape(): String {
```

Comment:
suggestion

This may be more efficient in time and concise if implemented with a map.  Here we have two iterations on the input string, once to check to see if anything should be done, and then again to do a conditional transform.  This could be collapsed to one iteration if there is a static `Map<Char, String>` which contains the escape values and then simply calling `Map.getOrElse(chr)` on each character.

Reply:
hmm ya I'll look into it